src/main/java/me/botsko/prism/actions/BlockAction.java

```diff
@@ -30,574 +30,534 @@
+     */
+    @Override
+    public ChangeResult applyDeferred(Player player, QueryParameters parameters, boolean isPreview) {
+        final Block block = getWorld().getBlockAt(getLoc());
+        return placeBlock(player, parameters, isPreview, block, true);
+    }
+
+    ChangeResult placeBlock(Player player, QueryParameters parameters, boolean isPreview, Block block,
+                            boolean isDeferred) {
+        BlockStateChange stateChange;
+
+        // Ensure block action is allowed to place a block here.
+        // (essentially liquid/air).
+
+        final boolean cancelIfBadPlace = !getActionType().requiresHandler(BlockChangeAction.class)
+                && !getActionType().requiresHandler(PrismRollbackAction.class) && !parameters.hasFlag(Flag.OVERWRITE);
+
+        if (cancelIfBadPlace && !BlockUtils.isAcceptableForBlockPlace(block.getType())) {
+            Prism.debug("Block skipped due to being unacceptable for block place.: " + block.getType().name());
+            return new ChangeResult(ChangeResultType.SKIPPED, null);
+        }
+
+        // On the blacklist (except an undo)
+        if (Prism.getIllegalBlocks().contains(getMaterial())
+                && !parameters.getProcessType().equals(PrismProcessType.UNDO)) {
+            Prism.debug("Block skipped because it's not allowed to be placed unless its an UNDO."
+                    + block.getType().name());
+            return new ChangeResult(ChangeResultType.SKIPPED, null);
+        }
+        // If we're not in a preview, actually apply this block
+        // Capture the block before we change it
+        final BlockState originalBlock = block.getState();
+        if (!isPreview) {
+            return handleApply(block, originalBlock, parameters, cancelIfBadPlace);
+        } else {
+
+            // Otherwise, save the state so we can cancel if needed
+            // Note: we save the original state as both old/new so we can re-use
+            // blockStateChanges
+            stateChange = new BlockStateChange(originalBlock, originalBlock);
+
+            // Preview it
+            EntityUtils.sendBlockChange(player, block.getLocation(), getBlockData());
+
+            // Send preview to shared players
+            for (final CommandSender sharedPlayer : parameters.getSharedPlayers()) {
+                if (sharedPlayer instanceof Player) {
+                    EntityUtils.sendBlockChange((Player) sharedPlayer, block.getLocation(), getBlockData());
+                }
+            }
+            return new ChangeResult(ChangeResultType.APPLIED, stateChange);
+        }
+    }
+
+    /**
+     * The BlockState object is modified by this method.
+     *
+     * @param block            Block
+     * @param originalBlock    BlockState
+     * @param parameters       QueryParameters
+     * @param cancelIfBadPlace cancelIfBadPlace
+     * @return ChangeResult.
+     */
+    private ChangeResult handleApply(final Block block, final BlockState originalBlock,
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [The method 'handleApply(Block, BlockState, QueryParameters, boolean)' has an NPath complexity of 193536](https://app.codacy.com/gh/AddstarMC/Prism-Bukkit/pullRequest?prid=5446892)